File: c-implementation/src/MemoryManager.c
#include "MemoryManager.h"

MemoryManager* create_memory_manager(int memory_size,
  int page_size, int swap_size){
  MemoryManager* mm = (MemoryManager*) malloc(sizeof(MemoryManager));
  
  // Assign frames and swap spaces in terms of page size.
  mm->num_frames = memory_size / page_size;
  mm->num_swap_pages = swap_size / page_size;

  // Assign frames and free frames - require num_frames array of ints.
  mm->frames = (int*) malloc(mm->num_frames * sizeof(int));
  mm->free_frames = (int*) malloc(mm->num_frames * sizeof(int));
  // On initialisation, all frames are free.
  mm->free_frame_count = mm->num_frames;

  // Set swap queue - require num_swap_pages array of ints
  mm->swap_queue = (int*) malloc(mm->num_swap_pages * sizeof(int));
  // At first, swap queue is empty
  mm->swap_queue_count = mm->num_swap_pages;

  // Assign memory to page table
  mm->page_table = (Page**) malloc(N_PAGES * sizeof(Page*));

  for (int i = 0; i < mm->num_frames; i++) {
    // No frames currently in use, se to -1
    mm->frames[i] = -1;
    // All frames are free, store indices in free_frames
    mm->free_frames[i] = i;
  }

  for (int i = 0; i < mm->num_swap_pages; i++) {
    // All slots in the swap queue initially free
    mm->swap_queue[i] = i;
  }

  for (int i = 0; i < 100; i++) {
    // Page table is empty as no pages loaded yet
    mm->page_table[i] = NULL;
  }

  return mm;
}

void access_page(MemoryManager *mm, int page_id){
  if (mm->page_table[page_id] == NULL){
    mm->page_table[page_id] = (Page*) malloc(sizeof(Page));
    mm->page_table[page_id]->id = page_id;
    mm->page_table[page_id]->in_memory = 0;
    // The page is no longer in a frame
    mm->page_table[page_id]->frame_number = -1;
  }

  Page *page = mm->page_table[page_id];

  // Handle case where page is already in memory
  if (page->in_memory) {
    printf("Page %d is already in memory at frame %d.\n", page_id, page->frame_number);
  } else {
    handle_page_fault(mm, page);
  }
}

void handle_page_fault(MemoryManager *mm, Page *page){
  printf("Page fault for page %d at address %p.\n", page->id, page);
  if (mm->free_frame_count > 0) {
    // If we have free frames, use them
    printf("Free frame count is %d.\n", mm->free_frame_count);
    int frame_number = mm->free_frames[--mm->free_frame_count];
    load_page_into_frame(mm, page, frame_number);
  } else {
    // If we have no free frames, swap out memory
    printf("Free frame count is %d, swapping memory.\n", mm->free_frame_count);
    swap_out_page(mm);
    int frame_number = mm->free_frames[--mm->free_frame_count];
    load_page_into_frame(mm, page, frame_number);
  }
}

void load_page_into_frame(MemoryManager *mm, Page *page, int frame_number){
  // Load page id into frame
  mm->frames[frame_number] = page->id;
  page->in_memory = 1;
  page->frame_number = frame_number;
  printf("Page %d at address %p is loaded into frame %d.\n", page->id, page, frame_number);
}

void swap_out_page(MemoryManager *mm){
  // Select a frame at random
  int frame_number = rand() % mm->num_frames;
  int page_id = mm->frames[frame_number];
  Page *page = mm->page_table[page_id];

  // Swapped page no longer in memory, written to 'disk'
  printf("Swapping out page %d at address %p from frame %d to swap space.\n", page->id, page, frame_number);
  page->in_memory = 0;
  page->frame_number = -1;

  // If the swap queue is not empty, put page in swap queue
  if (mm->swap_queue_count > 0) {
      mm->swap_queue[--mm->swap_queue_count] = page_id;
  }
  mm->free_frames[mm->free_frame_count++] = frame_number;
}

File: python-implementation/memory_manager.py
import random

class Page:

    def __init__(self, id):
        # Unique ID of page
        self.id = id
        # Whether of not page is in memory
        self.in_memory = False
        # Frame number in memory 
        self.frame_number = -1

class MemoryManager:

    DEFAULT_PAGE_SIZE = 4  # Page size in KB
    DEFAULT_MEMORY_SIZE = 16  # Physical memory size in KB
    DEFAULT_SWAP_SIZE = 32  # Swap space size in KB

    def __init__(self, memory_size=DEFAULT_MEMORY_SIZE, page_size=DEFAULT_PAGE_SIZE, swap_size=DEFAULT_SWAP_SIZE):
        # Assign frames and swap spaces in terms of page size.
        self.num_frames = memory_size // page_size
        self.num_swap_pages = swap_size // page_size
        self.frames = [-1] * self.num_frames
        self.free_frames = list(range(self.num_frames))
        self.swap_queue = list(range(self.num_swap_pages))

        # Initialise empty hash tables for pages and sswaps
        self.page_table = {}
        self.swap_space = {}

    def access_page(self, page_id):
        # Save page to page table if not originally there
        if page_id not in self.page_table:
            self.page_table[page_id] = Page(page_id)

        page = self.page_table[page_id]

        # Handle page faults from when page is already in emory.
        if page.in_memory:
            print(f"Page {page_id} is already in memory at frame {page.frame_number}.")
        else:
            self.handle_page_fault(page)

    def handle_page_fault(self, page):
        print(f"Page fault for page {page.id}.")
        if self.free_frames:
            # If we have free frames, use them
            frame_number = self.free_frames.pop(0)
            self.load_page_into_frame(page, frame_number)
        else:
            # If we don't have free frames, swap the memory out
            self.swap_out_page()
            frame_number = self.free_frames.pop(0)
            self.load_page_into_frame(page, frame_number)

    def load_page_into_frame(self, page, frame_number):
        # Load page id into specific frame
        print(f"Loading page {page.id} into frame {frame_number}.")
        self.frames[frame_number] = page.id
        page.in_memory = True
        page.frame_number = frame_number

    def swap_out_page(self):
        # Select a frame at random to swap iut - in practice
        # the least recently used frame will be chosen
        frame_number = random.randint(0, self.num_frames - 1)
        page_id = self.frames[frame_number]
        page = self.page_table[page_id]

        # Swapped page is no longer in memory
        print(f"Swapping out page {page.id} from frame {frame_number} to swap space.")
        page.in_memory = False
        page.frame_number = -1
        if self.swap_queue:
            self.swap_queue.pop(0)
        self.free_frames.append(frame_number)

def main():

    # Create memory manager
    memory_manager = MemoryManager()

    # Access pages. We only have four pages.
    memory_manager.access_page(1)
    memory_manager.access_page(2)
    memory_manager.access_page(3)
    memory_manager.access_page(4)
    memory_manager.access_page(5)  # This will cause a page fault and a swap since memory is full

if __name__ == "__main__":
    main()

